### PR TITLE
Multi GPU (WIP)

### PR DESCRIPTION
**Pull request by nakosung**
Multi GPU implementation(Multi tower, averaging gradient)

### CODE REVIEW

**Comment by nakosung** [comment]
I will squash commits after review completed. :)

**Comment by lemonzi** [comment]
@nakosung GitHub takes care of squashing during merge, don't worry about that :)

File: wavenet/model.py [hunk 1]
```diff
@@ -3,21 +3,36 @@
 from .ops import causal_conv, mu_law_encode
 
 
-def create_variable(name, shape):
+def _create_variable(name, shape):
```
Comment (jyegerlehner) [inline wavenet/model.py]: I thought the underscore was for member functions (methods) of classes. Am I wrong?

File: train.py [hunk 2]
```diff
@@ -214,38 +241,52 @@ def main():
...
+            average_grads.append(grad_and_var)
+        optim = optimizer.apply_gradients(average_grads)
 
     # Set up logging for TensorBoard.
     writer = tf.train.SummaryWriter(logdir)
     writer.add_graph(tf.get_default_graph())
     run_metadata = tf.RunMetadata()
-    summaries = tf.merge_all_summaries()
+    summaries = tf.merge_summary(summaries)
+    #summaries = tf.merge_all_summaries()
```
Comment (jyegerlehner) [inline train.py]: delete commented-out line

**Comment by astanway** [comment]
FWIW, I've tried running this branch on an Amazon p2.8xlarge (with 8 GPUs) and I am finding that while all GPUs log full memory utilization, the `volatile gpu-util` from `nvidia-smi` stays at 0% across all of them and the training time per step is the same as with a single GPU (~3 seconds).

**Comment by nakosung** [comment]
@astanway Actually this branch does additional minibatch, so you should tweak your learning rate and batch size to determine whether this branch works or not.

**Comment by akademi4eg** [comment]
This PR is marked as WIP. Are there any things not finished yet? (besides conflicts resolving)
Also, have anyone had a chance to test this PR on multi GPU setup? @astanway  @nakosung

File: wavenet/model.py [hunk 3]
```diff
@@ -3,21 +3,36 @@
...
 
-def create_bias_variable(name, shape):
+def _create_bias_variable(name, shape):
     '''Create a bias variable with the specified name and shape and initialize
     it to zero.'''
     initializer = tf.constant_initializer(value=0.0, dtype=tf.float32)
     return tf.Variable(initializer(shape=shape), name)
 
 
+def _get_variable(name, shape):
```
Comment (weixsong) [inline wavenet/model.py]: is the create_embedding_table() function should be changed to tf.get_variable() too? I'm not quite sure, if the embedding table is also trained as model parameters, should it be made shared among GPU?

Comment (nakosung) [inline wavenet/model.py]: @weixsong Yes. This PR is out-dated. :)

**Comment by weixsong** [comment]
Hi @nakosung , did you have some model that trained by this multiple GPU code? Is the results the same good as single GPU?

**Comment by nakosung** [comment]
@weixsong I haven't generated samples with multi-node version, but I think it should not differ from single-node version because its loss seemed good enough.
